core/server/src/main/java/alluxio/worker/block/BlockLockManager.java

```diff
@@ -122,16 +127,27 @@ public long lockBlock(long sessionId, long blockId, BlockLockType blockLockType)
    * @return the block lock
    */
   private ClientRWLock getBlockLock(long blockId) {
-    ClientRWLock blockLock;
-    synchronized (mSharedMapsLock) {
-      blockLock = mLocks.get(blockId);
-      if (blockLock != null) {
-        blockLock.addReference();
+    ClientRWLock blockLock = null;
+    boolean acquiredNewLock = false;
+    while (blockLock == null) {
+      // Check whether a lock has already been allocated for the block id.
+      synchronized (mSharedMapsLock) {
+        blockLock = mLocks.get(blockId);
+        if (blockLock != null) {
+          blockLock.addReference();
+        }
       }
-    }
-    if (blockLock == null) {
+      // If a block lock hasn't already been allocated, try to acquire a new one from the pool.
       // Acquire the lock outside the synchronized section because #acquire might need to block.
-      blockLock = mLockPool.acquire();
+      // We shouldn't wait indefinitely in acquire because the another lock for this block could be
+      // allocated to another thread, in which case we could just use that lock.
+      if (blockLock == null) {
+        blockLock = mLockPool.acquire(1, TimeUnit.SECONDS);
+        acquiredNewLock = (blockLock != null);
+      }
+    }
```

Comment:
Is there a race condition? ie. the lock is held from line 145 to 146 and freed afterward?